File: Test-Corpus/claude-no-features/B02_synthetic/pointer-comparison-ascii-art/translated_rust_original/c_src/src/scene.c
/*
 * Copyright 2025 MIT Lincoln Laboratory
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software
 * and associated documentation files (the "Software"),
 * to deal in the Software without restriction,
 * including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice
 * shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
 * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
// scene.c
#include "scene.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

scene_t* scene_create(const char *name) {
    scene_t *scene = malloc(sizeof(scene_t));
    if (!scene) {
        return NULL;
    }
    
    if (name) {
        strncpy(scene->name, name, MAX_SCENE_NAME - 1);
        scene->name[MAX_SCENE_NAME - 1] = '\0';
    } else {
        strcpy(scene->name, "Untitled Scene");
    }
    
    scene->shape_count = 0;
    
    return scene;
}

void scene_destroy(scene_t *scene) {
    if (scene) {
        // Note: We don't free the shapes themselves
        // because they are singletons managed by shape_manager
        free(scene);
    }
}

int scene_add_shape(scene_t *scene, shape_t *shape) {
    if (!scene || !shape) {
        return -1;
    }
    
    if (scene->shape_count >= MAX_SHAPES_IN_SCENE) {
        fprintf(stderr, "Error: Scene is full\n");
        return -1;
    }
    
    scene->shapes[scene->shape_count++] = shape;
    return 0;
}

int scene_remove_shape(scene_t *scene, int index) {
    if (!scene || index < 0 || index >= scene->shape_count) {
        return -1;
    }
    
    // Shift remaining shapes
    for (int i = index; i < scene->shape_count - 1; i++) {
        scene->shapes[i] = scene->shapes[i + 1];
    }
    
    scene->shape_count--;
    return 0;
}

void scene_print(const scene_t *scene) {
    if (!scene) {
        printf("(null scene)\n");
        return;
    }
    
    printf("\n=== Scene: %s ===\n", scene->name);
    printf("Contains %d shape(s)\n\n", scene->shape_count);
    
    for (int i = 0; i < scene->shape_count; i++) {
        printf("Shape #%d:\n", i + 1);
        shape_print(scene->shapes[i]);
        printf("\n");
    }
}

int scene_equals(const scene_t *s1, const scene_t *s2) {
    if (!s1 || !s2) {
        return 0;
    }
    
    // Scenes are equal if there's a 1:1 correspondence
    if (s1->shape_count != s2->shape_count) {
        return 0;
    }
    
    // For each shape in s1, find a matching shape in s2
    int matched[MAX_SHAPES_IN_SCENE] = {0};
    
    for (int i = 0; i < s1->shape_count; i++) {
        int found = 0;
        for (int j = 0; j < s2->shape_count; j++) {
            if (!matched[j] && shape_equals(s1->shapes[i], s2->shapes[j])) {
                matched[j] = 1;
                found = 1;
                break;
            }
        }
        if (!found) {
            return 0;
        }
    }
    
    return 1;
}

int scene_save(const scene_t *scene, const char *filename) {
    if (!scene || !filename) {
        return -1;
    }
    
    FILE *file = fopen(filename, "w");
    if (!file) {
        fprintf(stderr, "Error: Could not open file '%s' for writing\n", filename);
        return -1;
    }
    
    // Write scene name
    fprintf(file, "%s\n", scene->name);
    
    // Write shape count
    fprintf(file, "%d\n", scene->shape_count);
    
    // Write shape types (not the shapes themselves, just their types)
    for (int i = 0; i < scene->shape_count; i++) {
        fprintf(file, "%d\n", scene->shapes[i]->type);
    }
    
    fclose(file);
    printf("Scene saved to '%s'\n", filename);
    return 0;
}

scene_t* scene_load(const char *filename) {
    if (!filename) {
        return NULL;
    }
    
    FILE *file = fopen(filename, "r");
    if (!file) {
        fprintf(stderr, "Error: Could not open file '%s' for reading\n", filename);
        return NULL;
    }
    
    char name[MAX_SCENE_NAME];
    if (fgets(name, MAX_SCENE_NAME, file) == NULL) {
        fclose(file);
        return NULL;
    }
    
    // Remove newline
    name[strcspn(name, "\n")] = 0;
    
    scene_t *scene = scene_create(name);
    if (!scene) {
        fclose(file);
        return NULL;
    }
    
    int shape_count;
    if (fscanf(file, "%d\n", &shape_count) != 1) {
        scene_destroy(scene);
        fclose(file);
        return NULL;
    }
    
    for (int i = 0; i < shape_count; i++) {
        int type;
        if (fscanf(file, "%d\n", &type) != 1) {
            scene_destroy(scene);
            fclose(file);
            return NULL;
        }
        
        shape_t *shape = shape_get((shape_type_t)type);
        if (shape) {
            scene_add_shape(scene, shape);
        }
    }
    
    fclose(file);
    printf("Scene loaded from '%s'\n", filename);
    return scene;
}

void scene_list_shapes(const scene_t *scene) {
    if (!scene) {
        printf("(null scene)\n");
        return;
    }
    
    printf("\nScene: %s\n", scene->name);
    printf("Shapes (%d):\n", scene->shape_count);
    
    for (int i = 0; i < scene->shape_count; i++) {
        printf("  %d. %s (ptr: %p)\n", i + 1, 
               scene->shapes[i]->name, 
               (void*)scene->shapes[i]);
    }
}

File: Test-Corpus/claude/B02_synthetic/pointer-comparison-ascii-art/verified/src/scene.rs
//! Translation of `scene.c` / `scene.h`.

use std::fs::File;
use std::io::Write;
use std::os::unix::ffi::OsStrExt;
use std::path::Path;

use crate::cio::{self, CReader, Out};
use crate::p;
use crate::shape::{fmt_ptr, shape_equals, ShapeManager, ShapeRef};

pub const MAX_SHAPES_IN_SCENE: usize = 50;
pub const MAX_SCENE_NAME: usize = 64;

pub struct Scene {
    /// `char name[MAX_SCENE_NAME]`, stored as the raw bytes of the C string.
    pub name: Vec<u8>,
    /// `shape_t *shapes[MAX_SHAPES_IN_SCENE]` + `shape_count`.
    pub shapes: Vec<ShapeRef>,
}

impl Scene {
    pub fn shape_count(&self) -> i32 {
        self.shapes.len() as i32
    }
}

/// `scene_create`
pub fn scene_create(name: Option<&[u8]>) -> Option<Scene> {
    let name = match name {
        Some(n) => cio::truncate_to_buffer(n, MAX_SCENE_NAME),
        None => b"Untitled Scene".to_vec(),
    };
    Some(Scene {
        name,
        shapes: Vec::new(),
    })
}

/// `scene_add_shape`
pub fn scene_add_shape(scene: &mut Scene, shape: Option<ShapeRef>) -> i32 {
    let shape = match shape {
        Some(s) => s,
        None => return -1,
    };

    if scene.shapes.len() >= MAX_SHAPES_IN_SCENE {
        cio::err_str("Error: Scene is full\n");
        return -1;
    }

    scene.shapes.push(shape);
    0
}

/// `scene_remove_shape`
pub fn scene_remove_shape(scene: &mut Scene, index: i32) -> i32 {
    if index < 0 || index >= scene.shape_count() {
        return -1;
    }

    scene.shapes.remove(index as usize);
    0
}

/// `scene_print`
pub fn scene_print(out: &mut Out, mgr: &ShapeManager, scene: &Scene) {
    out.s("\n=== Scene: ");
    out.b(&scene.name);
    out.s(" ===\n");
    p!(out, "Contains {} shape(s)\n\n", scene.shape_count());

    for i in 0..scene.shape_count() {
        p!(out, "Shape #{}:\n", i + 1);
        mgr.print(out, Some(scene.shapes[i as usize]));
        out.s("\n");
    }
}

/// `scene_equals`
pub fn scene_equals(s1: &Scene, s2: &Scene) -> i32 {
    if s1.shape_count() != s2.shape_count() {
        return 0;
    }

    let mut matched = [0i32; MAX_SHAPES_IN_SCENE];

    for i in 0..s1.shape_count() as usize {
        let mut found = 0;
        for j in 0..s2.shape_count() as usize {
            if matched[j] == 0 && shape_equals(s1.shapes[i], s2.shapes[j]) != 0 {
                matched[j] = 1;
                found = 1;
                break;
            }
        }
        if found == 0 {
            return 0;
        }
    }

    1
}

/// `scene_save`
pub fn scene_save(out: &mut Out, mgr: &ShapeManager, scene: &Scene, filename: &[u8]) -> i32 {
    let path = Path::new(std::ffi::OsStr::from_bytes(filename));
    let file = match File::create(path) {
        Ok(f) => f,
        Err(_) => {
            cio::err_bytes(&[
                b"Error: Could not open file '",
                filename,
                b"' for writing\n",
            ]);
            return -1;
        }
    };
    let mut file = std::io::BufWriter::new(file);

    // Write scene name
    let _ = file.write_all(&scene.name);
    let _ = file.write_all(b"\n");

    // Write shape count
    let _ = file.write_all(format!("{}\n", scene.shape_count()).as_bytes());

    // Write shape types (not the shapes themselves, just their types)
    for i in 0..scene.shape_count() as usize {
        let kind = mgr.shape(scene.shapes[i]).kind;
        let _ = file.write_all(format!("{}\n", kind).as_bytes());
    }

    let _ = file.flush();
    drop(file);

    out.s("Scene saved to '");
    out.b(filename);
    out.s("'\n");
    0
}

/// `scene_load`
pub fn scene_load(out: &mut Out, mgr: &ShapeManager, filename: &[u8]) -> Option<Scene> {
    let path = Path::new(std::ffi::OsStr::from_bytes(filename));
    let file = match File::open(path) {
        Ok(f) => f,
        Err(_) => {
            cio::err_bytes(&[
                b"Error: Could not open file '",
                filename,
                b"' for reading\n",
            ]);
            return None;
        }
    };
    let mut file = CReader::new(file);

    let name = match file.fgets(MAX_SCENE_NAME) {
        Some(line) => line,
        None => return None,
    };

    // Remove newline
    let name = cio::strip_newline(&name);

    let mut scene = match scene_create(Some(&name)) {
        Some(s) => s,
        None => return None,
    };

    let shape_count = match file.scan_int() {
        Some(v) => {
            file.skip_format_space();
            v
        }
        None => return None,
    };

    for _ in 0..shape_count {
        let shape_type = match file.scan_int() {
            Some(v) => {
                file.skip_format_space();
                v
            }
            None => return None,
        };

        let shape = mgr.get(shape_type);
        if shape.is_some() {
            scene_add_shape(&mut scene, shape);
        }
    }

    out.s("Scene loaded from '");
    out.b(filename);
    out.s("'\n");
    Some(scene)
}

/// `scene_list_shapes`
pub fn scene_list_shapes(out: &mut Out, mgr: &ShapeManager, scene: &Scene) {
    out.s("\nScene: ");
    out.b(&scene.name);
    out.s("\n");
    p!(out, "Shapes ({}):\n", scene.shape_count());

    for i in 0..scene.shape_count() as usize {
        let r = scene.shapes[i];
        p!(
            out,
            "  {}. {} (ptr: {})\n",
            i + 1,
            mgr.name(r),
            fmt_ptr(mgr.addr(r))
        );
    }
}
